python/dgl/contrib/sampling/sampler.py

```diff
@@ -77,10 +84,124 @@ def __next__(self):
             aux_infos['seeds'] = self._seed_ids.pop(0).tousertensor()
         return self._subgraphs.pop(0), aux_infos
 
+class _Prefetcher(object):
+    """Internal shared prefetcher logic. It can be sub-classed by a Thread-based implementation
+    or Process-based implementation."""
+    _dataq = None  # Data queue transmits prefetched elements
+    _controlq = None  # Control queue to instruct thread / process shutdown
+    _errorq = None  # Error queue to transmit exceptions from worker to master
+
+    _checked_start = False  # True once startup has been checkd by _check_start
+
+    def __init__(self, loader, num_prefetch):
+        super(_Prefetcher, self).__init__()
+        self.loader = loader
+        assert num_prefetch > 0, 'Unbounded Prefetcher is unsupported.'
+        self.num_prefetch = num_prefetch
+
+    def run(self):
+        """Method representing the process’s activity."""
+        # Startup - Master waits for this
+        try:
+            loader_iter = iter(self.loader)
+            self._errorq.put(None)
+        except Exception as e:  # pylint: disable=broad-except
+            tb = traceback.format_exc()
+            self._errorq.put((e, tb))
+
+        while True:
+            try:  # Check control queue
+                c = self._controlq.get(False)
+                if c is None:
+                    break
+                else:
+                    raise RuntimeError('Got unexpected control code {}'.format(repr(c)))
+            except queue.Empty:
+                pass
+            except RuntimeError as e:
+                tb = traceback.format_exc()
+                self._errorq.put((e, tb))
+                self._dataq.put(None)
+
+            try:
+                data = next(loader_iter)
+                error = None
+            except Exception as e:  # pylint: disable=broad-except
+                tb = traceback.format_exc()
+                error = (e, tb)
+                data = None
+            finally:
+                self._errorq.put(error)
+                self._dataq.put(data)
+
+    def __next__(self):
+        next_item = self._dataq.get()
+        next_error = self._errorq.get()
+
+        if next_error is None:
+            return next_item
+        else:
+            self._controlq.put(None)
+            if isinstance(next_error[0], StopIteration):
+                raise StopIteration
+            else:
+                return self._reraise(*next_error)
+
+    def _reraise(self, e, tb):
+        print('Reraising exception from Prefetcher', file=sys.stderr)
+        print(tb, file=sys.stderr)
+        raise e
+
+    def _check_start(self):
+        assert not self._checked_start
+        self._checked_start = True
+        next_error = self._errorq.get(block=True)
```

Comment:
why do we need to block here?

Reply:
This makes sure the `iter` method is called in the worker thread. If there's any error fetching the first batch, the error will be reported during initialization